### DISCUSSION TITLE
Self-Compiled SbieDrv.sys different than WHCP-signed version

### DISCUSSION THREAD

**Discussion by Ailbeenlynd** [sandboxie-plus/Sandboxie#4214] [General]
I was able to use VS2019 community edition to compile Sandboxie Classic from source code using all default settings. I then disassembled my self-compiled SbieDrv.sys and the WHCP-signed SbieDrv.sys from the same release and compared the two. For earlier versions of Sandboxie (e.g., right after the Sophos hand over), the two SbieDrv.sys were usually roughly identical with only minor differences, however, for later versions, the WHCP-signed SbieDrv.sys always contains a few more unknown functions than the my self-compiled file. Why is that? I understand that different compiler/linker settings could contribute to that, if that's the case, can those settings be shared? Ideally, it'd be good to see both the self-compiled SbieDrv.sys and the downloaded/WHCP-signed version be identical, so we can safely use the WHCP-signed driver. I am not saying those extraneous functions are illegitimate, but would like to understand how they got into the driver that's supposedly built from the same code. thanks

**Comment by DavidXanatos** [comment]
The driver that gets signed is whats produced by the CI build mechanism run by github, you can pick the corresponding build artifacts from https://github.com/sandboxie-plus/Sandboxie/actions all binaries from the right action should be identical to the signed once, except ofcause for the attached signature.

**Reply by Ailbeenlynd** [reply]
When built with CI, the two matched. All's good, thanks for your hard work!

**Comment by Ailbeenlynd** [comment]
Sorry, all is not good yet... you added driver enforced signature verification. We don't have your privatekey, I can't just use your WHCP-signed SbieDrv.sys without also using your Start.exe, SbieCtrl.exe and SbieSvc.exe, is that right? I guess I can't disable signature verification in the driver either, since then I will lose the WHCP signature. This additional protection is a bit too restrictive for those wanting to self-compile but can't afford to WHCP sign their driver.

**Reply by DavidXanatos** [reply]
Thank you for your feedback. I understand the challenges with the driver signature enforcement. Here are three possible options that don't require WHCP signing:

* Disable Driver Signature Enforcement
You can place your system into Test Mode to bypass the driver signature requirement. This allows you to run self-compiled drivers without WHCP signing.

* Use a Custom Kernel Signer
There are tools available to help you use custom kernel signatures, which can help circumvent the need for official signing:

[Geoff Chappell's Custom Kernel Signers Guide](https://www.geoffchappell.com/notes/windows/license/customkernelsigners.htm)
[Windows 10 Custom Kernel Signers on GitHub](https://github.com/HyperSine/Windows10-CustomKernelSigners)
[SSDE on GitHub](https://github.com/valinet/ssde)

* Develop in a VM in Test Mode

Alternatively, you can develop and test your code in a virtual machine running in Test Mode. Once everything is stable, commit the changes to the main project and wait for the next official signed release.

I hope one of these solutions works for your situation!

**Reply by Ailbeenlynd** [reply]
Thank you. Does SbieDrv.sys only enforce signature for SbieSvc.exe or all 3 .exe's with .sig? I am testing on an older version, it seems to work with the signed SbieSvc.exe with SbieSvc.exe.sig alone (without the signed Start.exe and SbieCtrl.exe).

**Reply by DavidXanatos** [reply]
all 3 exec, but its mostly critical only for SbieSvc.exe, 
start.exe is mostly not affected
sandman.exe and sbiectrl.exe must be signed to be able to controll process excemptions, that is clicking allow on a allow network prompt or alike, and to set itself as session leader.

I could add an option to a specific certificate to disable this signature enforcement, making it a developer certificate, would that be of interest?

**Reply by Ailbeenlynd** [reply]
How would that work? Do you mean you would add exceptions to the standard release SbieDrv.sys to bypass signature enforcement if the .exe's are signed with a specific developer certificate? This would create a security loophole if the developer certificate is made publicly available? I must be missing something

**Reply by DavidXanatos** [reply]
No a user could install a special developer certificate instead of a regular supporter certificate which would for their PC disable the signature enforcement, and only their PC the certificate could be generated to be bound to their specific HwID, like business use certs are.

**Reply by Ailbeenlynd** [reply]
That makes sense. I don't think I need that level of exception for my usage scenario yet, maybe wait to see if others are interested? I will look into the custom kernel signer route you referenced, thanks for the offer and always being responsive!

**Comment by love-code-yeyixiao** [comment]
In a word,you may not use self-complier driver in a completely normal environment.




---Original---
From: ***@***.***&gt;
Date: Sun, Sep 8, 2024 02:28 AM
To: ***@***.***&gt;;
Cc: ***@***.***&gt;;
Subject: Re: [sandboxie-plus/Sandboxie] Self-Compiled SbieDrv.sys differentthan WHCP-signed version (Discussion #4214)




 
That makes sense. I don't think I need that level of exception for my usage scenario yet, maybe wait to see if others are interested? I will look into the custom kernel signer route you referenced, thanks for the offer and always being responsive!
